File: FinalProjSubmission/HS_Output/applyOp.cpp
:
<|im_end|>
<|im_start|>assistant
```cpp
#include <iostream>

enum class Op {
  Add,
  Sub,
  Mul
};

int applyOp(Op op, int x, int y) {
  switch (op) {
    case Op::Add:
      return x + y;
    case Op::Sub:
      return x - y;
    case Op::Mul:
      return x * y;
    default:
      throw std::invalid_argument("Invalid operation");
  }
}

int evalList(Op op, const std::vector<int>& xs) {
  int result = 0;
  for (int x : xs) {
    result = applyOp(op, result, x);
  }
  return result;
}

int main() {
  std::cout << evalList(Op::Add, {1, 2, 3, 4, 5}) << std::endl;
  std::cout << evalList(Op::Sub, {5, 1, 2}) << std::endl;
  std::cout << evalList(Op::Mul, {2, 3, 4}) << std::endl;
  return 0;
}
```<|im_end|>

File: FinalProjSubmission/HS_Dataset/applyOp.hs
import Text.Printf (printf)
data Op = Add | Sub | Mul

applyOp :: Op -> Int -> Int -> Int
applyOp Add x y = x + y
applyOp Sub x y = x - y
applyOp Mul x y = x * y

evalList :: Op -> [Int] -> Int
evalList _ [] = 0
evalList op (x:xs) = applyOp op x (evalList op xs)

main = do
    printf "%d\n" $ evalList Add [1,2,3,4,5]
    printf "%d\n" $ evalList Sub [5,1,2]
    printf "%d\n" $ evalList Mul [2,3,4]